keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Multinomial.java

```diff
@@ -0,0 +1,134 @@
+        return new Multinomial(n, p);
+    }
+
+    /**
+     * https://en.wikipedia.org/wiki/Multinomial_distribution
+     * Generalisation of the Binomial distribution to variables with >2 possible values
+     *
+     * @param n The number of draws from the variable
+     * @param p The probability of observing each of the k values (which sum to 1)
+     *          p is a Tensor whose first dimension must be of size k
+     */
+    private Multinomial(IntegerTensor n, DoubleTensor p) {
+        Preconditions.checkArgument(
+            p.sum(0).elementwiseEquals(DoubleTensor.ones(n.getShape())).allTrue(),
+            "Probabilities must sum to one"
+        );
+
+        numCategories = p.getShape()[0];
+        for (int category = 0; category < numCategories; category++) {
+            TensorShapeValidation.checkAllShapesMatch(n.getShape(), p.slice(0, category).getShape());
+        }
+        this.n = n;
+        this.p = p;
+    }
+
+    @Override
+    public IntegerTensor sample(int[] shape, KeanuRandom random) {
+        TensorShapeValidation.checkTensorsMatchNonScalarShapeOrAreScalar(shape, n.getShape());
+
+        p.sum(0);
+        Tensor.FlattenedView<Integer> nFlattened = n.getFlattenedView();
+        List<Tensor.FlattenedView<Double>> pFlattened = Lists.newArrayList();
+
+        for (int category = 0; category < numCategories; category++) {
+            DoubleTensor p_i = p.slice(0, category);
+            pFlattened.add(p_i.getFlattenedView());
+        }
+
+        int length = ArrayUtil.prod(shape);
+        int[] samples = new int[0];
+
+        for (int i = 0; i < length; i++) {
+            final int j = i;
+            List<Double> ps = pFlattened.stream().map(p -> p.getOrScalar(j)).collect(Collectors.toList());
+            int[] sample = sample(nFlattened.getOrScalar(i), random, ps.toArray(new Double[0]));
+            samples = ArrayUtils.addAll(samples, sample);
+        }
+        int[] outputShape = shape;
+        if (shape[0] == 1) {
+            outputShape = ArrayUtils.remove(outputShape, 0);
+        }
+        outputShape = ArrayUtils.add(outputShape, numCategories);
+        return IntegerTensor.create(samples, outputShape).transpose();
+
+    }
+
+    private static int[] sample(int n, KeanuRandom random, Double... p) {
+        int[] counts = new int[p.length];
+        for (int i = 0; i < n; i++) {
+            int index = sample(random, p);
+            counts[index] += 1;
+        }
+        return counts;
+    }
+
+    private static int sample(KeanuRandom random, Double... p) {
+        double value = random.nextDouble();
+        int index = 0;
+        Double pCumulative = 0.;
+        while (index < p.length) {
+            Double currentP = p[index++];
+            if (currentP == 0.) {
+                continue;
+            }
+            pCumulative += currentP;
+            if (pCumulative >= value) {
+                break;
+            }
+        }
+        return index-1;
```

Comment:
Is sampling -1 an expected behaviour?

Reply:
That can only happen when `p.length` is zero. I think we can reasonably assume that will never happen and it doesn't require a unit test.